### ISSUE TITLE
Allow for client-side reloads with `pserve --reload`

### ISSUE THREAD

**Issue by cdunklau** [Pylons/pyramid#3352]
It would be super nice if there was a way to trigger a *client-side* reload after `pserve --reload` respawns the server. This kind of thing is quite common in JavaScript tooling, to the point that's included in some project skeletons (e.g. create-react-app). The right way to implement this... well that's trickier than just wishing for it! 😃

My rough idea is to teach `pserve` to invoke a callback after the reload completes. I suspect that would be a good 80% solution, and it offers a lot of flexibility.

I'm working with Grunt and [grunt-contrib-watch](https://github.com/gruntjs/grunt-contrib-watch), which uses [tiny-lr](https://github.com/mklabs/tiny-lr) for its livereload support. My post-pserve-reload callback would just [POST /changed](https://github.com/mklabs/tiny-lr#tiny-lr-post-changed) to the livereload server, and that'd be the end of it.

I'll be playing with this idea a bit before I can throw together a PR for more focused discussion, but please comment with suggestions.

**Comment by mmerickel** [comment]
`hupper` restarts the entire process when it reloads your app. You can do things when your app is reloaded basically anywhere, for example in your `main` or at module-scope or whatever else. See below for some example code to use to detect the situation.

As far as waiting until your app is actually ready to serve requests, there is no standard hook for that and is dependent on your WSGI server. The best approach, even if that hook existed, would be to deal with it client-side by retrying until it can actually connect before telling livereload to reload.

```python
def get_hupper_reloader():
    import hupper
    try:
        return hupper.get_reloader():
    except Exception:
        pass

reloader = get_hupper_reloader()
if reloader is not None:
    # I'm running in hupper - do things
```

**Comment by cdunklau** [comment]
Thanks @mmerickel. I didn't look closely enough at what pserve was doing... since it's actually calling the WSGI server instead of spawning a child to run it, I can see how such a hook wouldn't be doable in pserve without a large rewrite.

With your pointers I've come up with a solution using Twisted's WSGI support. Here's how I did it, in case it's useful to anyone else. It's likely brittle in a lot of ways, especially around errors involving the livereload API, but it works for me for the time being.

To run, just the normal `pserve --reload development.ini` does the trick.

The `server` section in my `development.ini` file looks like this:

```ini
[server:main]
use = call:devserver:main
listen = localhost:6543
livereload.server_port = 35729
```

And finally, `devserver.py` has this (Python 3.6+):

```python3
import json

from twisted.web.wsgi import WSGIResource
from twisted.web import client, server
from twisted.web.http_headers import Headers
from twisted.application import service, strports
from twisted.python import threadpool
from twisted.internet import defer
from twisted.logger import Logger, STDLibLogObserver, globalLogBeginner
from twisted.web.iweb import IBodyProducer
from zope.interface import implementer


try:
    import hupper
except ImportError:
    hupper = None


log = Logger()


def makeWSGIResource(reactor, wsgiApp):
    # Adapted from twisted.plugins.twisted_web
    pool = threadpool.ThreadPool()
    reactor.callWhenRunning(pool.start)
    reactor.addSystemEventTrigger("after", "shutdown", pool.stop)
    resource = WSGIResource(reactor, pool, wsgiApp)
    return resource


class LiveReloadNotifier(object):
    def __init__(self, agent, liveReloadPort):
        self.agent = agent
        self.target = f"http://localhost:{liveReloadPort}/changed"

        self._dfd = None

    def notifyChanged(self):
        if self._dfd is not None:
            raise Exception(
                "LiveReloadNotifier callback called more than once!"
            )
        log.info("Triggering live reload")
        self._dfd = req = self.agent.request(
            b"POST",
            self.target.encode("ascii"),
            Headers({"Content-Type": ["application/json"]}),
            # We have to send _something_ for files, otherwise the API
            # won't trigger a page reload.
            JSONBodyProducer({"files": [""]}),
        )

        def cbCheckReponse(response):
            if response.code != 200:
                raise UnexpectedLiveReloadResponse(
                    f"Response code {response.code} != 200"
                )
            return response

        req.addCallback(cbCheckReponse)
        req.addCallback(client.readBody)
        req.addCallback(lambda body: json.loads(body.decode("utf-8")))
        req.addCallback(
            lambda structure: log.info(
                "Success, got structure: {structure}", structure=str(structure)
            )
        )
        req.addErrback(
            lambda f: log.failure(
                "LiveReload notification request failed", failure=f
            )
        )


@implementer(IBodyProducer)
class JSONBodyProducer:
    def __init__(self, structure):
        self.data = json.dumps(structure).encode("utf-8")
        self.length = len(self.data)

    def startProducing(self, consumer):
        consumer.write(self.data)
        return defer.succeed(None)

    def stopProducing(self):
        pass

    def pauseProducing(self):
        pass

    def resumeProducing(self):
        pass


def makeWebApplication(rootResource, endpointDescription):
    webApplication = service.Application("devserver")
    site = server.Site(rootResource)
    servColl = service.IServiceCollection(webApplication)
    s = strports.service(endpointDescription, site)
    s.setServiceParent(servColl)
    return webApplication


class UnexpectedLiveReloadResponse(Exception):
    pass


def run(reactor, wsgiApp, *, listen, reloaderPort=None):
    listenIface, _, listenPort = listen.partition(":")
    endpointDescription = f"tcp:{listenPort}:interface={listenIface}"

    rootResource = makeWSGIResource(reactor, wsgiApp)
    webApplication = makeWebApplication(rootResource, endpointDescription)

    if reloaderPort is not None:
        agent = client.Agent(reactor)
        notifier = LiveReloadNotifier(agent, reloaderPort)

        # FIXME: This should be when it's ready to receive requests...
        # but there's probably some more proper way via the Twisted
        # application framework.
        reactor.addSystemEventTrigger(
            "after", "startup", notifier.notifyChanged
        )

    service.IService(webApplication).startService()
    reactor.run()


def runningUnderReloader():
    if hupper is None:
        return False
    try:
        hupper.get_reloader()
    except RuntimeError:
        return False

    return True


def main(global_config, **settings):
    config_spec = [
        ("listen", "listen", "localhost:6543", None),
        ("reloaderPort", "livereload.server_port", "35729", int),
    ]
    runKeywords = {}
    for kwargName, settingsKey, default, coercer in config_spec:
        value = settings.get(settingsKey, default)
        if coercer is not None:
            value = coercer(value)
        runKeywords[kwargName] = value

    if not runningUnderReloader():
        runKeywords.pop("reloaderPort")

    def serveWSGI(wsgiApp):
        from twisted.internet import reactor

        observers = [STDLibLogObserver()]
        globalLogBeginner.beginLoggingTo(observers)

        run(reactor, wsgiApp, **runKeywords)

    return serveWSGI
```

**Comment by stevepiercy** [comment]
@cdunklau I'd accept a PR to the Pyramid Community Cookbook. Perhaps a new page under the category [Development Tools](https://docs.pylonsproject.org/projects/pyramid-cookbook/en/latest/development_tools/index.html) or another new category or Miscellaneous? There's also a [Pyramid/WebPack setup for collecting static assets](https://docs.pylonsproject.org/projects/pyramid_cookbook/en/latest/static_assets/bundling-static-assets.html). I think it would be cool to extend it with client auto-reload, so you don't have to switch to a browser and manually reload it, especially when working with forms.

**Comment by cdunklau** [comment]
@stevepiercy good idea about the cookbook, but I don't think the hacky solution with Twisted is a good idea for it... It turns out I went on a wild goose chase. Using a task runner that properly handles the child process workflow should be enough to use pserve/waitress like in the usual documentation examples.

This originally came about because I was having trouble with grunt-contrib-watch and pserve (without reload), such that Grunt would restart pserve when a watched file changed. The issue was that when I changed a watched file, Grunt's restart of pserve resulted in "Address already in use" errors. I made a bad assumption that this was a fault with waitress, like it wasn't using SO_REUSEADDR or something, so I went down the Twisted route.

I then noticed the startup time of my Twisted solution was substantially longer, so I dug in a bit more, and found out that pserve (without reload) + waitress in fact works fine with fast terminate/relaunch. I get no errors with this code, it restarts fine:

```python3
import time
import subprocess
import contextlib

@contextlib.contextmanager
def terminating(proc):
    try:
        yield proc
    finally:
        proc.terminate()
        proc.wait()


cmd = ['venv/bin/pserve', 'development.ini']
with terminating(subprocess.Popen(cmd)) as proc:
    time.sleep(5)
with terminating(subprocess.Popen(cmd)) as proc:
    time.sleep(5)
```

It turns out that my issue seems to be caused by gruntjs/grunt-contrib-watch#484
